src/lib/Util.hs

```diff
@@ -241,3 +242,22 @@ measureSeconds m = do
   ans <- m
   t2 <- liftIO $ getCPUTime
   return (ans, (fromIntegral $ t2 - t1) / 1e12)
+
+-- === zippable class ===
+
+class Zippable f where
+  zipWithZ :: MonadFail m => (a -> b -> m c) -> f a -> f b -> m (f c)
+
+zipWithZ_ :: Zippable f => MonadFail m => (a -> b -> m c) -> f a -> f b -> m ()
+zipWithZ_ f xs ys = zipWithZ f xs ys >> return ()
+
+zipErr :: MonadFail m => m a
+zipErr = fail "zip error"
+
+forMZipped :: MonadFail m => [a] -> [b] -> (a -> b -> m c) -> m [c]
+forMZipped xs ys f
```

Comment:
If we defined `Zippable` for lists, wouldn't that give us `forMZipped` through `zipWithZ`?

Reply:
Yep, good point